packages/toolkit/src/query/core/buildMiddleware/windowEventHandling.ts

```diff
@@ -35,24 +39,27 @@ export const build: SubMiddlewareBuilder = ({
         const querySubState = queries[queryCacheKey]
         const subscriptionSubState = subscriptions[queryCacheKey]
 
-        if (
-          !subscriptionSubState ||
-          !querySubState ||
-          querySubState.status === QueryStatus.uninitialized
-        )
-          return
-
-        const shouldRefetch =
-          Object.values(subscriptionSubState).some(
-            (sub) => sub[type] === true
-          ) ||
-          (Object.values(subscriptionSubState).every(
-            (sub) => sub[type] === undefined
-          ) &&
-            state.config[type])
-
-        if (shouldRefetch) {
-          api.dispatch(refetchQuery(querySubState, queryCacheKey))
+        if (!subscriptionSubState || !querySubState) return
+
+        if (Object.keys(subscriptionSubState).length === 0) {
+          api.dispatch(
+            removeQueryResult({
+              queryCacheKey: queryCacheKey as QueryCacheKey,
+            })
+          )
+        } else if (querySubState.status !== QueryStatus.uninitialized) {
+          const shouldRefetch =
+            Object.values(subscriptionSubState).some(
+              (sub) => sub[type] === true
+            ) ||
+            (Object.values(subscriptionSubState).every(
+              (sub) => sub[type] === undefined
+            ) &&
+              state.config[type])
+
+          if (shouldRefetch) {
+            api.dispatch(refetchQuery(querySubState, queryCacheKey))
+          }
```

Comment:
Should `removeQueryResult` be guarded by `shouldRefetch` as well? What do you think @phryneas ?

Reply:
Yes, moving it into that guard might be a good way. That way we remove as little as necessary from the cache.